internals/controllers/migrationhierarchy/migrationhierarchy_controller.go

```diff
@@ -0,0 +1,440 @@
+	if err := sns.EnsureDeleteObject(); err != nil {
+		reason := fmt.Sprintf("deleting the subnamespace failed because: " + err.Error())
+		return reason, err
+	}
+
+	return "", nil
+}
+
+// updateRole updates the role of a subnamespace
+func (r *MigrationHierarchyReconciler) updateRole(sns *utils.ObjectContext, role string) error {
+	err := sns.UpdateObject(func(object client.Object, l logr.Logger) (client.Object, logr.Logger) {
+		object.(*corev1.Namespace).GetLabels()[danav1.Role] = role
+		return object, l
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// enqueueOriginalParent enqueues the original parent of the migrated subnamespace for reconciliation
+// so that its status is updated after one or more of its children get migrated
+func (r *MigrationHierarchyReconciler) enqueueOriginalParent(ctx context.Context, sourceSNSParentName string) error {
+	sourceSNSParentNS, err := utils.NewObjectContext(ctx, r.Client, client.ObjectKey{Namespace: "", Name: sourceSNSParentName}, &corev1.Namespace{})
+	if err != nil {
+		return fmt.Errorf("failed getting namespace object '%s': "+err.Error(), sourceSNSParentName)
+	}
+
+	sourceSnsParentNamespace := utils.GetNamespaceParent(sourceSNSParentNS.Object)
+	r.addSnsToSnsEvent(sourceSNSParentName, sourceSnsParentNamespace)
+
+	return nil
+}
+
+// enqueueSNSDescendants enqueues the descendants of a subnamespace for reconciliation
+func (r *MigrationHierarchyReconciler) enqueueSNSDescendants(sns *utils.ObjectContext) {
+	snsDescendants := utils.GetAllChildren(sns)
+	for _, sns := range snsDescendants {
+		r.addSnsToSnsEvent(sns.Object.GetName(), sns.Object.GetNamespace())
+	}
+}
+
+// addSnsToSnsEvent takes two parameters: snsName and snsNamespace
+// then adds the sns to the sns event channel to trigger new re-sync for the sns
+func (r *MigrationHierarchyReconciler) addSnsToSnsEvent(snsName string, snsNamespace string) {
+	r.SnsEvents <- event.GenericEvent{Object: &danav1.Subnamespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      snsName,
+			Namespace: snsNamespace,
+		},
+	}}
+}
+
+// UpdateNSBasedOnParent updates the labels and annotations of a namespace
+// based on its parent labels and annotations
+func (r *MigrationHierarchyReconciler) UpdateNSBasedOnParent(ctx context.Context, parentNS, childNS *utils.ObjectContext) error {
+	nsName := childNS.Object.GetName()
+	labels, annotations := utils.GetNSLabelsAnnotationsBasedOnParent(parentNS, nsName)
+
+	if err := childNS.AppendAnnotations(annotations); err != nil {
+		return err
+	}
+
+	if err := childNS.AppendLabels(labels); err != nil {
+		return err
+	}
+
+	// update the ClusterResourceQuota AnnotationSelector if needed
+	isChildNSResourcePool, err := utils.IsNamespaceResourcePool(childNS)
+	if err != nil {
+		return err
+	}
+
+	isChildNSUpperResourcePool, err := utils.IsNSUpperResourcePool(childNS)
+	if err != nil {
+		return err
+	}
+
+	if !isChildNSResourcePool || isChildNSUpperResourcePool {
```

Comment:
this section is not undetstandble wrap it in function and function will explain what it does

Reply:
wrapped